### PR TITLE
Create unique string id for each BERT backbone

### LINKED ISSUE #363: Create unique string id for each BERT pretrained checkpoint
Currently BERT pretrained encoders are specified by their vocabulary alone. This is possible because checkpoints must be called from the correct class (e.g., `BertBase('uncased_en')`).

However, if we want to request checkpoints from any combination of architectures and vocabularies in a task-specific model (e.g., `BertClassifier`) it will be much easier if each checkpoint has a unique name.

Other benefits:
* Will also make it easy to document all of our checkpoints in keras.io. There are many possible checkpoints that could be specified by `'uncased_en'` but only one by `'bert_base_uncased_en'`.
* Currently the `backbone` model can only be specified by instantiating a Bert model. This change allows string ids and defaults for the backbone models.

### PR DESCRIPTION
**Pull request by jbischof**
Closes #363

### CODE REVIEW

**Comment by jbischof** [comment]
@mattdangerw actually we could make this more maintainable by moving `backbone_checkpoints` to `bert.py` and making it the source of truth. Then the current checkpoints struct would have the sole job of telling us the which checkpoints work with each class (e.g., 'uncased_en' for `BertBase` is in `backbone_checkpoints['bert_base_uncased_en']`).

**Comment by jbischof** [comment]
I am confident I can clean up and consolidate the structs in `bert.py`, but mainly we need to decide if we want unique checkpoint ids or the shortest strings possible.
```python
# Not unique, but short as possible
BertBase('uncased_en')
# Unique but not short
BertBase('bert_base_uncased_en')
```

File: keras_nlp/models/bert_tasks.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+
+class BertClassifier(keras.Model):
+    """BERT encoder model with a classification head.
+
+    Args:
+        backbone: A string, `keras_nlp.models.BertCustom` or derivative such as
+            `keras_nlp.models.BertBase` to encode inputs. If a string, should be
+            one of `keras_nlp.models.bert_tasks.backbone_checkpoints`.
```
Comment (mattdangerw) [inline keras_nlp/models/bert_tasks.py:103]: So this is something a user would print out? I think we might want to think about how we export this (and other such lists for preprocessing and backbone models).

We probably should explicitly choose where to export and document these symbols through `__init__.py` rather than couple our file structure. We also may want to add a `bert` directory, so we can control the exports in `keras_nlp.models.bert` via an `__init__.py`. No need to figure all that out on this PR though.

For now, should we just do the same format string with dict keys approach we did for the other bert symbols?

File: keras_nlp/network_tests/test_load_ckpts.py [hunk 2]
```diff
@@ -44,6 +44,26 @@ def test_load_bert(self, bert_variant, weights):
         }
         model(input_data)
 
+    def test_load_bert_backbone_string(self):
+        for (
```
Comment (mattdangerw) [inline keras_nlp/network_tests/test_load_ckpts.py:48]: split this into two lines for readability

```
checkpoint_names = keras_nlp.models.bert_tasks.backbone_checkpoints.keys()
for checkpoint in checkpoin_names:
```
```

File: keras_nlp/network_tests/test_load_ckpts.py [hunk 3]
```diff
@@ -44,6 +44,26 @@ def test_load_bert(self, bert_variant, weights):
         }
         model(input_data)
 
+    def test_load_bert_backbone_string(self):
```
Comment (mattdangerw) [inline keras_nlp/network_tests/test_load_ckpts.py:47]: we have a dict of keys we could traverse for all of these tests. should we just do this approach everywhere rather than a parameterized test? seems much more maintainable

**Review by mattdangerw — COMMENTED** [review]
Looks good! Leaving some quick comments on the PR as I ran through it. But the main question you called out re-checkpoint naming seems key. Will think and reply directly there.

**Comment by mattdangerw** [comment]
Re the question of unique vs shortest strings, it seems like both could work, with very interesting tradeoffs.

But it seems incomplete without considering `BertPreprocessor`. Unique identifiers for checkpoints, but shortest possible identifiers for preprocessing feels like we would be asking users to mix two different mental models.

The shortest possible string proposal with preprocessing:
```
BertPreprocessor('uncased_en')
BertBase('uncased_en')
BertClassifier(backbone='base_uncased_en')

# And down the line...
BertClassifier('sst_base_uncase_en')
```

The unique ID proposal, also applied to preprocessing:
```
BertPreprocessor('bert_base_uncased_en')
BertBase('bert_base_uncased_en')
BertClassifier(backbone='bert_base_uncased_en')

# And down the line...
BertPreprocessor('bert_sst_base_uncased_en')
BertClassifier('bert_sst_base_uncased_en')
```

It feels like a non-obvious choice. Honestly the main reason to go with an ID approach would seems to be thinking about this in a far off future state, where we want to handle tons of checkpoints, or even integrating with some (or many!) external hubs. I'm not really sure how to evaluate that.

The former option seems the most in line with Keras today. Not trying to say that should trump other considerations though.

**Comment by jbischof** [comment]
Excellent point @mattdangerw! I think the short answer is that the preprocessor construction needs to work from either a model name prompt or direct vocab prompt. The user may want to train their own model using a pretrained vocab or want us to find the right vocab for a checkpoint.

```python
# Direct request for vocab
BertPreprocessor(vocabulary='uncased_en')
# Request to for correct vocab for a checkpoint
BertPreprocessor(weights='bert_based_uncased_en')
# Especially helpful for task driven models where vocab may be obvious
BertPreprocess(weights="ToxicBERT")
```

Since using `weights` is a bit confusing in this context (is this the checkpoint for a vocab or the vocab for a checkpoint?), we could also allow either to be specified as `vocabulary`. A model will basically never have the same name as a vocab.

```python
# Direct request for vocab
BertPreprocessor(vocabulary='uncased_en')
# Request to for correct vocab for a checkpoint
BertPreprocessor(vocabulary='bert_based_uncased_en')
# Especially helpful for task driven models where vocab may be obvious
BertPreprocess(vocabulary="ToxicBERT")
```

Under the hood we first look in the `vocabularies` struct and then `checkpoints` one. Now a user can never misunderstand...we accept both! WDYT?

File: keras_nlp/models/bert_tasks.py (hunk 1 above)
Comment (jbischof) [reply keras_nlp/models/bert_tasks.py:103]: Agree on the `bert/` directory, although I may file an issue and reogranize `models/` in a followup PR.

I was worried about the large number of checkpoints and long names that the formatting would be horrible. Is there any way to prevent the docstring from having one 1000 char line? I of course would have made the checkpoint names a string-backed `enum` and called it a day!

**Comment by mattdangerw** [comment]
In the ID world (if we go with that), why does `'uncased_en'` have to mean anything anymore? I'm not sure it does. You could either provide a checkpoint ID, or you provide a vocab yourself.

```python
BertPreprocessor(vocabulary='bert_base_uncased_en')
or
BertPreprocessor(vocabulary='./my-cool-vocab.txt')
```

Worry with your proposal above is two fold I guess...

1) The whole point of fully specified string name would be to give the name universal meaning throughout all symbols we offer. That alone is sufficient right? Why also introduce another namespace just for specifying vocabularies directly? We never need to support a preset vocabulary that is not in that universal ID namespace.
2) `weights` current in Keras applications is supposed to either be a "name shortcut" or a literal weights file (tf or h5 format). The analog for `vocabulary` would seem to be, either a "name shortcut" or a literal vocab file. It seems most consistent to keep that approach. If we want to diverge to avoid overloaded args, it seems better to do that with a totally different name. `BertPreprocessor(preset='bert_base_uncased_en')`, `BertBase(preset='bert_base_uncased_en')` or `BertPreprocessor(vocabulary='./some-vocab.txt')`, `BertPreprocessor(weights='./some-weights.h5')`. Using `weights` in the way you are proposing above seems like it is both inconsistent with the rest of Keras, and breaking the user mental model for the name (i.e. that `weights` refers to some literal floats somewhere).

**Comment by jbischof** [comment]
Sounds good @mattdangerw let's figure out the preprocessor in a followup PR! We were just sanity checking how unique string ids work for vocabs.

File: keras_nlp/network_tests/test_load_ckpts.py [hunk 4]
```diff
@@ -15,41 +15,63 @@
...
+                    shape=(1, 512), dtype=tf.int64, maxval=model.vocabulary_size
+                ),
+                "segment_ids": tf.constant(
+                    [0] * 200 + [1] * 312, shape=(1, 512)
+                ),
+                "padding_mask": tf.constant([1] * 512, shape=(1, 512)),
+            }
+            model(input_data)
+
+    def test_load_bert_backbone_string(self):
```
Comment (mattdangerw) [inline keras_nlp/network_tests/test_load_ckpts.py:43]: maybe test_load_bert_backbone_string -> test_classifier_backbone_string

just for consistency with name below

File: keras_nlp/models/bert_tasks.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    )
+    logits = classifier(input_data)
+    ```
+"""
+
+
+class BertClassifier(keras.Model):
+    def __init__(
+        self,
+        backbone="bert_base_uncased_en",
```
Comment (mattdangerw) [inline keras_nlp/models/bert_tasks.py:74]: Seems like if we make this change here we should also make the changes to add a default for the weights argument right? That feels like it would be most consistent.

We can do that as a follow up though.

File: keras_nlp/models/bert.py [hunk 6]
```diff
@@ -782,41 +734,50 @@ def BertLarge(weights=None, vocabulary_size=None, name=None, trainable=True):
     return model
 
 
+arch_classes = {
```
Comment (mattdangerw) [inline keras_nlp/models/bert.py:737]: This feels hard to read name everywhere it's imported. Seeing just the import I have no idea what an arch_class is.

Can we just inline this in a function? `lookup_model_class(name)` or whatever we want the name to be. Could also do this as a follow up.

**Review by mattdangerw — APPROVED** [review]
Thank you!

This looks good to me if we have overall consensus on the "checkpoint id" approach.

We should followup with making preprocessing also use checkpoint ids, and give defaults to the `weights` argument for the `BertBase` and similar classes.

File: keras_nlp/models/bert_tasks.py (hunk 5 above)
Comment (jbischof) [reply keras_nlp/models/bert_tasks.py:74]: Do you mean have a default for every `weights` argument in `bert.py`?

File: keras_nlp/network_tests/test_load_ckpts.py (hunk 4 above)
Comment (jbischof) [reply keras_nlp/network_tests/test_load_ckpts.py:43]: I was unsure if we would repeat this for every task type. I was eventually planning to extract this functionality into a helper method (added a TODO).

File: keras_nlp/models/bert_tasks.py (hunk 5 above)
Comment (mattdangerw) [reply keras_nlp/models/bert_tasks.py:74]: Yep! That seems consistent to me at least. If `BertClassifier()` gets you an uncased en backbone, and `BertPreprocessor()` gets you an uncased en vocabulary, it seems weird that `BertBase()` gets you an error message.

Let's be consistent with our choice of defaults. Either we choose a nice default everywhere (and keep in mind for some models like GPT-2 there might be a clear choice), or we ask users to be explicit everywhere.

And totally fine to do this as a follow up!

Comment (jbischof) [reply keras_nlp/models/bert_tasks.py:74]: Filed [#370](https://github.com/keras-team/keras-nlp/issues/370)

File: keras_nlp/models/bert_tasks.py (hunk 1 above)
Comment (jbischof) [reply keras_nlp/models/bert_tasks.py]: Fixed for now by adding all checkpoints to the docstring. For the longer term added a TODO in `bert.py` about using keras.io to programatically document checkpoint coverage.

File: keras_nlp/models/bert_tasks.py (hunk 5 above)
Comment (mattdangerw) [reply keras_nlp/models/bert_tasks.py:74]: Danke! I'll add some more color there on my two cents.

**Comment by jbischof** [comment]
Got offline approval from @fchollet
